### PR TITLE
Ygot diff tool

### PR DESCRIPTION
**Pull request by sachinholla**
#### Why I did it

Subscription enhancements proposed in https://github.com/sonic-net/SONiC/pull/1287 requires gNMI server to create a SAMPLE notification message by comparing YGOT objects from two successive iterations. We observed that the standard `ygot.Diff` API takes more time and memory when the YGOT objects have multiple child objects. Hence developed an in-house diff API

#### How I did it

Performance issues of `ygot.Diff` API are mainly due to it flattening the YGOT objects into list of {leaf_path, leaf_value} pairs and then comparing those lists. The in-house `Diff` API avoids this flattening step. It traverses both YGOT object attributes together, compares the values inline and records the changes if any. Following table captures the time taken (in nanoseconds) by `ygot.Diff` and the in-house diff APIs for diffing two openconfig-acl objects. Size "MxN" stands for ygot object with M ACLs and N rules in each ACL. All values are in nanoseconds.

| Test case         | API           | size=10x10  | size=10x100  | size=100x1000 |
|-------------------|---------------|------------:|-------------:|--------------:|
| No changes        | ygot.Diff     |  9024089697 | 854023452896 |           OOM |
|                   | in-house Diff |     1784778 |     15460520 |    1757874254 |
| One attr changed  | ygot.Diff     |  9285911808 | 857292949475 |           OOM |
|                   | in-house Diff |     1547512 |     16506496 |    1786245108 |
| All attrs changed | ygot.Diff     |  9333680743 | 855581319202 |          OOM  |
|                   | in-house Diff |     3562498 |     34306626 |    3846524858 |
| Create all attrs  | ygot.Diff     |    80830018 |    858752644 |               |
|                   | in-house Diff |     7101530 |     74663888 |    8245685970 |
| Delete all attrs  | ygot.Diff     |    83082154 |    755762682 |               |
|                   | in-house Diff |      120232 |       125293 |        453306 |

#### How to verify it

Tested the functionality through gotests that handcraft two ygot objects, invoke the new Diff function and verify the results.
The gNMI server code to use this API is not ready as of now.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: transl_utils/ygot_diff.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+			v2 := iter.Value()
+			yd.pushPrefix(listName, &v2)
+			yd.recordUpdates(v2.Interface())
+			yd.popPrefix()
+		}
+	}
+}
+
+func (yd *ygotDiff) forUnion(f *reflect.StructField, v1, v2 *reflect.Value) {
+	if v2.IsNil() {
```
[the remark this answers is not in the document]
Comment (sachinholla) [reply transl_utils/ygot_diff.go:144]: `forUnion()` will be called only when the value is of `reflect.Interface` kind -- line 105.

File: transl_utils/ygot_diff.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+		pElem.Key, err = ygot.PathKeyFromStruct(*keyObj)
+		if err != nil {
+			panic("ygot.PathKeyFromStruct failed: " + err.Error())
+		}
+	}
+	return pElem
+}
+
+func unboxPtr(v reflect.Value) interface{} {
+	if v.IsNil() {
```
[the remark this answers is not in the document]
Comment (sachinholla) [reply transl_utils/ygot_diff.go:332]: `unboxPtr()` is called only when the value of a `reflect.Ptr` kind

File: transl_utils/ygot_diff.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	if suffix != nil {
+		elems[len(yd.prefix)] = suffix
+	}
+
+	return &gnmi.Path{Elem: elems}
+}
+
+// pushPrefix appends a new path element to the prefix.
+func (yd *ygotDiff) pushPrefix(name string, keyObj *reflect.Value) {
+	yd.prefix = append(yd.prefix, newPathElem(name, keyObj))
```
[the remark this answers is not in the document]
Comment (sachinholla) [reply transl_utils/ygot_diff.go:296]: I removed the unnecessary empty name check from `newPathElem()`. Name can never be empty there.
Thanks

**Review by sallylsy — APPROVED** [review]

**Review by tomek-US — APPROVED** [review]
LGTM

**Review by anand-kumar-subramanian — APPROVED** [review]

**Pull request merged**